src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpBindJoinSPARQLwithUNIONTest.java

```diff
@@ -0,0 +1,75 @@
+package se.liu.ida.hefquin.engine.queryplan.executable.impl.ops;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import org.apache.jena.graph.Graph;
+import org.apache.jena.sparql.core.Var;
+import org.junit.Test;
+
+import se.liu.ida.hefquin.engine.federation.SPARQLEndpoint;
+import se.liu.ida.hefquin.engine.query.TriplePattern;
+import se.liu.ida.hefquin.engine.queryplan.ExpectedVariables;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpTPAdd;
+import se.liu.ida.hefquin.engine.queryplan.physical.impl.PhysicalOpBindJoinWithUNION;
+
+public class ExecOpBindJoinSPARQLwithUNIONTest extends TestsForTPAddAlgorithms<SPARQLEndpoint> {
+	
+	@Test
+	public void tpWithJoinOnObject() {
+		_tpWithJoinOnObject();
+	}
+
+	@Test
+	public void tpWithJoinOnSubjectAndObject() {
+		_tpWithJoinOnSubjectAndObject();
+	}
+
+	@Test
+	public void tpWithoutJoinVariable() {
+		_tpWithoutJoinVariable();
+	}
+
+	@Test
+	public void tpWithEmptyInput() {
+		_tpWithEmptyInput();
+	}
+
+	@Test
+	public void tpWithEmptySolutionMappingAsInput() {
+		_tpWithEmptySolutionMappingAsInput();
+	}
+
+	@Test
+	public void tpWithEmptyResponses() {
+		_tpWithEmptyResponses();
+	}
+
+	@Test
+	public void tpWithSpuriousDuplicates() {
+		_tpWithSpuriousDuplicates();
+	}
+
+	@Override
+	protected SPARQLEndpoint createFedMemberForTest( final Graph dataForMember) {
+		return new SPARQLEndpointForTest(dataForMember);
+	}
+
+	@Override
+	protected UnaryExecutableOp createExecOpForTest( final TriplePattern tp, final SPARQLEndpoint fm) {
+		final LogicalOpTPAdd tpadd = new LogicalOpTPAdd(fm, tp);
+		final PhysicalOpBindJoinWithUNION bindjoin = new PhysicalOpBindJoinWithUNION(tpadd); 
+		return bindjoin.createExecOp(bindjoin.getExpectedVariables(new ExpectedVariables() {
```

Comment:
Here I create a new `PhysicalOpBindJoinWithUNION` object in order to call `createExecOp`, which receives a list of `ExpectedVariables`. To do that I call the `getExpectedVariables` method of the logical Op. In turn, this method also receives a list of `ExpectedVariables`, which I'm defining as empty.

For all tests, the possible variables are always empty. Even in the spurious duplicates test that I just created. So I'm inclined to believe that the condition that needs to be checked is another (it can also be that my intuition of the expected variables here is wrong). But I can't figure out how to check it anywhere else than in running time in the `fetchSolutionMappings` method.

Reply:
What you do here makes no sense. Usually (i.e., during our query planning process), the `ExpectedVariables` object that is passed to the `createExecOp` method is the `ExpectedVariables` object from the child operator and not from the operator itself (see the class `QueryPlanCompilerImpl` which is where this happens). Now, for unit tests, it should be an `ExpectedVariables` object that reflects the expected variables (certain and possible) of the solution mappings that the unit test will later on send as input to the `ExecOpBindJoinSPARQLwithUNION`.

Generally, the purpose of using `ExpectedVariables` objects during query planning is precisely that certain things do *not* need to be checked anymore at running time. For instance, if variable ?x is in the certain variables in the `ExpectedVariables` object that is passed to a `createExecOp` method, then the executable operator created by that `createExecOp` method can assume that *every* solution mapping that will later be sent to it as input is guaranteed to contain a binding for ?x. Similarly, if variable ?y is in the possible variables in the `ExpectedVariables` object that is passed to a `createExecOp` method, then the executable operator created by that `createExecOp` method can assume that *some* solution mappings that will later be sent to it as input have a binding for ?y; it may then turn out at runtime that every input solution mapping has a binding for ?y, but there is no guarantee for this.